lightningd/pay.c

```diff
@@ -624,54 +634,76 @@ bool wait_payment(const tal_t *cxt,
 	return cb_not_called;
 }
 
+static struct custom_route_data route_to_custom_route_data(
```

Comment:
it's fairly unconventional to return structs as values from my experience, usually you would just pass in a pointer.

Reply:
Well, we do pass back quite a few pointers to structs if we pass in a context we can allocate off.

Reply:
right that would work too

Reply:
I see the point you make about the layering violation, and it still isn't resolved in my later commits. Another thing I don't like about my approach is that this data gets tal_stolen when it gets passed to send_payment_custom_route_data. So I'm open for suggestions on how to improve this.

However, I'm not sure I fully understand what you suggest. Maybe it's easiest if you just write the code improvement yourself?

Side note: I was wondering how well tal combines with data allocated on the stack. I guess that when a stack-allocated object is used as a context, the child objects are not automatically going to be deallocated when the context object on the stack goes out of scope, right? This isn't C++ after all.